### ISSUE TITLE
Add Rule Version Filtering Based on @SinceKtlint Annotations

### ISSUE THREAD

**Issue by hugoncosta** [pinterest/ktlint#3099]
# Enhancement: Add Rule Version Filtering Based on @SinceKtlint Annotations

## Expected Behavior

Add a new configuration option to KtLint plugin that allows users to filter out rules that were introduced after a specified KtLint version. This would use the existing `@SinceKtlint` annotations present on all rules to determine which rules to include.

**Proposed Configuration:**
```kotlin
ktlint {
    version.set("1.5.0")
    maxRuleVersion.set("0.46.0") // Only include rules introduced up to version 0.46.0
}
```

Why do I want this?
Upgrading to higher versions of Kotlin (e.g. 2.1.0) requires a newer version of Ktlint (we're currently on 1.2.1). If I upgrade it, I will cause failures due to the new rules not being respected, which I'd like to avoid. 

With the proposed feature, I would be able to upgrade to the latest version (e.g. 1.7.1) which _does_ have Kotlin 2.2 compatibility without having to introduce all the new rules between 1.2.1 and 1.7.1. This would then allow me to upgrade the rule set on my own time.

## Current Behavior

No filtering allowed unless I manually exclude new rules myself.

## Additional Information

* Current version of ktlint: all

**Comment by paul-dingemans** [comment]
This feature request should be submitted to the KtLint Gradle project. That project is a wrapper around the KtLint project. I doubt it will work what you request. Your problem is related to the KtLintRuleEngine which is using a newer version of the embedded Kotlin compiler, and not on the rules themselves.

From ktlint perspective there is nothing that I can do to help you mitigating the problem.

**Comment by hugoncosta** [comment]
Yes, there are currently Kotlin compatibility issues that can probably be addressed in the Ktlint Gradle plugin project. I would still very much like to have the ability to be running on the latest version of Ktlint but _only_ bring in the rules up to a particular version. 

Internally, we centrally configure Ktlint for all projects, and we cannot be bump the version without causing breakages across the company. This feature would allow us to set a baseline version (we upgrade it for every major Gradle version) and let the customers bump it manually if they want to be on the latest and greatest.

**Comment by paul-dingemans** [comment]
I understand what you are aiming at. But the proposed configuration block would need to be supported by the KtLint-Gradle plugin.

Also, it is the Ktlint API Consumer, e.g. the Ktlint-Gradle plugin, that provides the rules that needs to be executed by the KtLintRuleEngine. The plugin can already  provide rules from older rulesets to the latest version of KtLintRuleEngine. For example the ktlint-intellij-plugin already works like this. Of course this only works when rulesets and KtLintRuleEngine have the same major version.

There is no need to filter rules by version when the Ktlint API Consumer just uses the correct dependency version of the rulesets.

**Comment by hugoncosta** [comment]
I see, thank you! I assumed both approaches would be valid, and if we were to do it within Ktlint itself, we wouldn't need a custom solution in every plugin. We can revisit this in the future. In the meantime, I'll take your advice and propose the feature in Ktlint Gradle itself. Appreciate your input!
